src/extendedstorage.cpp

```diff
@@ -626,18 +629,53 @@ void ExtendedStorage::Private::resetAlarms(const Incidence::List &incidences,
     setAlarms(incidences, calendar);
 }
 
+QDateTime ExtendedStorage::Private::getNextOccurrence(const Incidence::Ptr &incidence,
+                                                      const QDateTime &start,
+                                                      const Incidence::List &exceptions)
+{
+    if (!start.isNull() && incidence->recurs()) {
+        Recurrence *recurrence = incidence->recurrence();
+        QSet<QDateTime> recurrenceIds;
+        for (const Incidence::Ptr &exception : exceptions)
+            recurrenceIds.insert(exception->recurrenceId());
+
+        QDateTime match = start;
+        if (!recurrence->recursAt(start) || recurrenceIds.contains(start)) {
+            do {
+                match = recurrence->getNextDateTime(match);
+            } while (match.isValid() && recurrenceIds.contains(match));
+        }
+        return match;
+    } else {
+        return incidence->dtStart();
+    }
+}
+
 void ExtendedStorage::Private::setAlarms(const Incidence::List &incidences,
                                          const Calendar::Ptr &calendar)
 {
     const QDateTime now = QDateTime::currentDateTime();
     Timed::Event::List events;
-    foreach (const Incidence::Ptr incidence, incidences) {
+    for (Incidence::Ptr incidence : incidences) {
         // The incidence from the list must be in the calendar and in a notebook.
         const QString &nbuid = calendar->notebook(incidence->uid());
         if (!calendar->isVisible(incidence) || nbuid.isEmpty()) {
             continue;
         }
-        setAlarms(incidence, nbuid, events, now);
+        if (incidence->recurs()) {
+            const QDateTime next = getNextOccurrence(incidence, now, calendar->instances(incidence));
+            addAlarms(incidence, nbuid, &events, next);
+        } else if (incidence->hasRecurrenceId()) {
+            const Incidence::Ptr parent = calendar->incidence(incidence->uid());
+            if (parent && !incidences.contains(parent)) {
```

Comment:
Hm, suppose contains() loops over all the items, being here inside a loop over all the items. Wonder if that can take time when applied to a whole notebook. QList would have gotten toSet() to provide faster lookup but Incidence::List appears to be QVector. What do you think?

Reply:
Yeh, rightful concern. Normal case should have few exceptions, so the N**2 loop should actually be a N one. But worst case will indeed be N**2. I've reworked a bit the routine to first list the recurringUids in a QSet and then look in this hashtable for matching parent or not.